### ISSUE TITLE
Slow down of solver after recent updates.

### ISSUE THREAD

**Issue by lhy11009** [geodynamics/aspect#4959]
Recently, I have observed a slowdown in computation speed in our models (@mibillen ). I looked into this and found the " Solve Stokes system" causing the slowdown, where the difference is about 1/3.  I would post it here and see if others have consistent results. This includes a test on 1 CPU and 512 CPUs.

For a case run on 1 CPU (output_1_2_1), the comparison between the current and a previous run from 2021 October. The results show the difference in the entry of "Solve Stokes system". Previously, it was 79s, which has increased to 123s recently. This causes the total wallclock to increase from a previous 115 s to 164s. Moreover, other cases also agree with these results.
Another case run on 4 nodes, 512 CPUs (output_512_5_1), consumed 576s in "Solve Stokes system" last October, and 849s recently. The total wallclock time increases from 670s to 953s.

(detail of wallclock for the case on 1 CPU)
Results from 2021 October:
+----------------------------------------------+------------+------------+
| Total wallclock time elapsed since start     |       115s |            |
|                                              |            |            |
| Section                          | no. calls |  wall time | % of total |
+----------------------------------+-----------+------------+------------+
| Assemble Stokes system           |         2 |      11.7s |        10% |
| Assemble composition system      |         2 |      3.53s |       3.1% |
| Assemble temperature system      |         2 |      6.13s |       5.3% |
| Build Stokes preconditioner      |         2 |      8.27s |       7.2% |
| Build composition preconditioner |         2 |     0.703s |      0.61% |
| Build temperature preconditioner |         2 |     0.706s |      0.61% |
| Initialization                   |         1 |     0.754s |      0.66% |
| Postprocessing                   |         2 |     0.376s |      0.33% |
| Setup dof systems                |         1 |     0.301s |      0.26% |
| Setup initial conditions         |         1 |     0.889s |      0.77% |
| Setup matrices                   |         1 |      1.64s |       1.4% |
| Solve Stokes system              |         2 |        79s |        69% |
| Solve composition system         |         2 |     0.106s |         0% |
| Solve temperature system         |         2 |    0.0983s |         0% |
+----------------------------------+-----------+------------+------------+

Results from a recent run:
+----------------------------------------------+------------+------------+
| Total wallclock time elapsed since start     |       164s |            |
|                                              |            |            |
| Section                          | no. calls |  wall time | % of total |
+----------------------------------+-----------+------------+------------+
| Assemble Stokes system           |         2 |      13.7s |       8.4% |
| Assemble composition system      |         2 |      3.18s |       1.9% |
| Assemble temperature system      |         2 |      4.68s |       2.9% |
| Build Stokes preconditioner      |         2 |      10.2s |       6.2% |
| Build composition preconditioner |         2 |     0.33jkk33807s |      0.49% |
| Build temperature preconditioner |         2 |     0.811s |       0.5% |
| Initialization                   |         1 |      1.48s |       0.9% |
| Postprocessing                   |         2 |     0.439s |      0.27% |
| Setup dof systems                |         1 |     0.272s |      0.17% |
| Setup initial conditions         |         1 |      1.13s |      0.69% |
| Setup matrices                   |         1 |      2.57s |       1.6% |
| Solve Stokes system              |         2 |       123s |        75% |
| Solve composition system         |         2 |     0.186s |      0.11% |
| Solve temperature system         |         2 |     0.172s |       0.1% |
+----------------------------------+-----------+------------+------------+

**Comment by tjhei** [comment]
What are the exact ASPECT and deal.II versions?

**Comment by lhy11009** [comment]
Sorry, I forget to mention. These are the versions.
-- This is ASPECT, the Advanced Solver for Problems in Earth's ConvecTion.
--     . version 2.5.0-pre (master_TwoD, 5a5f21131)
--     . using deal.II 9.4.0
--     .       with 32 bit indices and vectorization level 2 (256 bits)
--     . using Trilinos 12.18.1
--     . using p4est 2.3.2
--     . running in OPTIMIZED mode
--     . running with 1 MPI process

**Comment by tjhei** [comment]
I only see a single ASPECT version. Are you saying things got slower with the same ASPECT version?

**Comment by lhy11009** [comment]
Oh, right, let me check the previous cases

**Comment by lhy11009** [comment]
-----------------------------------------------------------------------------
-- This is ASPECT, the Advanced Solver for Problems in Earth's ConvecTion.
--     . version 2.4.0-pre (main, 875a17cf8)
--     . using deal.II 9.3.1
--     .       with 32 bit indices and vectorization level 1 (128 bits)
--     . using Trilinos 13.0.0
--     . using p4est 2.8.0
--     . running in OPTIMIZED mode
--     . running with 128 MPI processes
-----------------------------------------------------------------------------

**Comment by lhy11009** [comment]
Sorry, sometimes I forget to include the full information.

**Comment by lhy11009** [comment]
Also, I suspect it's just our server. I have some clues here, from the affinity test

**Comment by lhy11009** [comment]
Here is the test with  Stampede2 "skx-normal" partition. This partition has 48 CPUs per core.
![Stampede](https://user-images.githubusercontent.com/28665365/191588831-b916d14c-1724-4ff8-b7f9-a052878a978a.png)
This is some differences in the header of the log file now that I double checked - the "vectorization" is shown to have different levels:
-----------------------------------------------------------------------------
-- This is ASPECT, the Advanced Solver for Problems in Earth's ConvecTion.
--     . version 2.5.0-pre (master_TwoD, f353920e1)
--     . using deal.II 9.4.0
--     .       with 32 bit indices and vectorization level 3 (512 bits)
--     . using Trilinos 12.18.1
--     . using p4est 2.3.2
--     . running in OPTIMIZED mode
--     . running with 2 MPI processes
-----------------------------------------------------------------------------

**Comment by gassmoeller** [comment]
@lhy11009 the difference in vectorization level can be very important, in particular if you use the GMG solver. It determines how good the vector units of the CPUs are used when solving the equations. The option is set when deal.II is compiled, and if you use candi it is controlled by the NATIVE_OPTIMIZATIONS option in candi.cfg (you should always set that to true, not sure if we have documented that @tjhei).
 
Also as Timo mentioned please attach the **complete** log.txt files (as files) for both the old and the new run (ideally also the `statistics` files and the .prm file if it is easy to run without additional data). Otherwise we cannot start to check why the timing is different.

**Comment by tjhei** [comment]
Yes, vectorization is required for good GMG performance. We document this here:
https://aspect-documentation.readthedocs.io/en/latest/user/methods/geometric-multigrid.html

I wonder if we should issue a warning if it is not enabled in release mode.

**Comment by lhy11009** [comment]
Okay, so that would be helpful indeed. However, I am sure that I am using the "block AMG" here all through.Here are the files.
New ones:
[log.txt](https://github.com/geodynamics/aspect/files/9620478/log.txt)
[original.prm.txt](https://github.com/geodynamics/aspect/files/9620509/original.prm.txt)
[statistics.txt](https://github.com/geodynamics/aspect/files/9620510/statistics.txt)
Here are the old ones:
[log.txt](https://github.com/geodynamics/aspect/files/9620522/log.txt)
[original.prm.txt](https://github.com/geodynamics/aspect/files/9620525/original.prm.txt)
[statistics.txt](https://github.com/geodynamics/aspect/files/9620526/statistics.txt)

**Comment by tjhei** [comment]
> I am using the "block AMG"

I have no idea how much of a difference native optimizations give for AMG, but I would be surprised if the difference is that big. More likely (in that order):
- machine issues
- Trilinos version
- Some regression in deal.II 4.0

**Comment by gassmoeller** [comment]
Hi @lhy11009 did you mean to close this issue? Can you quickly summarize your conclusion? (I am just trying to figure out if it is still relevant for #5136).

**Comment by gassmoeller** [comment]
Oh, I see you mentioned you conclusion in #5136, I should have checked that first. No need to reply here as well.
